File: src/agent_sessions/homefree/mux.py
"""A tiny stream multiplexer for the Home Free full-app tunnel (#579 P1).

The E2E ``Transport`` (handshake.py) is a single ordered, reliable, encrypted
byte channel. To carry the box's whole app — many concurrent ``/api`` requests
plus the terminal ``/ws`` — over that one channel, we multiplex independent
**streams** through it.

The mux is deliberately **transport-agnostic**: it emits opaque frame bytes via
an ``on_send`` callback and consumes them via :meth:`Mux.feed`. The caller wires
those to ``transport.encrypt`` → relay WebSocket and back, so the relay still
only ever sees ciphertext (no relay change). That also makes the mux testable by
feeding two instances straight into each other.

Wire — one frame is ``stream_id(u32) · type(u8) · len(u24) · payload[len]``:

======  ====  =====================================================
type    code  payload
======  ====  =====================================================
OPEN    1     opaque open-info (caller-defined; e.g. an HTTP request line)
DATA    2     stream bytes (subject to the per-stream send window)
END     3     empty — half-close (no more DATA from this sender)
RESET   4     1-byte error code — abort the stream both directions
WINDOW  5     u32 credit increment — the peer may send this many more bytes
======  ====  =====================================================

Flow control is credit-based per stream: a receiver advertises
``INITIAL_WINDOW`` and only replenishes (via ``WINDOW``) as its consumer reads,
so an un-read stream drains the sender's window and applies real backpressure
instead of buffering without bound. Stream ids follow the HTTP/2 convention —
the initiator opens odd ids, the responder even — so the two sides never collide.
"""

from __future__ import annotations

import asyncio
from collections import deque

OPEN, DATA, END, RESET, WINDOW = 1, 2, 3, 4, 5

INITIAL_WINDOW = 256 * 1024  # per-stream receive credit advertised to the peer
MAX_CHUNK = 16 * 1024  # DATA payload cap — keep frames small for fair interleaving
_U24_MAX = (1 << 24) - 1


def _encode(stream_id: int, ftype: int, payload: bytes = b"") -> bytes:
    if len(payload) > _U24_MAX:
        raise ValueError("mux frame payload too large")
    return stream_id.to_bytes(4, "big") + bytes([ftype]) + len(payload).to_bytes(3, "big") + payload


def _decode(frame: bytes) -> tuple[int, int, bytes]:
    if len(frame) < 8:
        raise ValueError("short mux frame")
    stream_id = int.from_bytes(frame[:4], "big")
    ftype = frame[4]
    length = int.from_bytes(frame[5:8], "big")
    # Require an exact length — reject short AND overlong frames, so trailing or
    # coalesced bytes can't be silently dropped (matches the TS decoder).
    if len(frame) != 8 + length:
        raise ValueError("mux frame length mismatch")
    return stream_id, ftype, frame[8 : 8 + length]


class StreamReset(Exception):
    """Raised on a stream whose peer sent RESET (or the mux was closed)."""

    def __init__(self, code: int = 0) -> None:
        super().__init__(f"stream reset (code {code})")
        self.code = code


class Stream:
    """One logical bidirectional channel over the mux."""

    def __init__(self, mux: Mux, stream_id: int, open_info: bytes = b"") -> None:
        self._mux = mux
        self.id = stream_id
        self.open_info = open_info
        # send side: how many bytes the peer will still accept
        self._send_window = INITIAL_WINDOW
        self._send_waiters: deque[asyncio.Future] = deque()
        self._send_ended = False
        # receive side: buffered bytes waiting for our consumer + EOF/reset state
        self._recv_buf = bytearray()
        self._recv_eof = False
        self._recv_reset: int | None = None
        self._data_ready = asyncio.Event()

    # ── outbound ──────────────────────────────────────────────────────
    async def write(self, data: bytes) -> None:
        """Send bytes, awaiting when the peer's window is exhausted (backpressure)."""
        if self._send_ended:
            raise StreamReset(0)
        view = memoryview(data)
        while view:
            while self._send_window <= 0:
                if self._recv_reset is not None:
                    raise StreamReset(self._recv_reset)
                fut = asyncio.get_event_loop().create_future()
                self._send_waiters.append(fut)
                await fut
            n = min(len(view), self._send_window, MAX_CHUNK)
            self._mux._emit(_encode(self.id, DATA, bytes(view[:n])))
            self._send_window -= n
            view = view[n:]

    async def end(self) -> None:
        """Half-close: no more outbound DATA."""
        if not self._send_ended:
            self._send_ended = True
            self._mux._emit(_encode(self.id, END))

    def reset(self, code: int = 0) -> None:
        """Abort the stream in both directions."""
        self._mux._emit(_encode(self.id, RESET, bytes([code & 0xFF])))
        self._mux._drop(self.id)
        self._fail(code)

    # ── inbound ───────────────────────────────────────────────────────
    async def read(self, max_bytes: int = MAX_CHUNK) -> bytes:
        """Return up to ``max_bytes`` of received data; ``b""`` at clean EOF.

        Reading is what replenishes the peer's send window — so a consumer that
        stops reading applies backpressure rather than letting us buffer forever.
        """
        while not self._recv_buf and not self._recv_eof and self._recv_reset is None:
            self._data_ready.clear()
            await self._data_ready.wait()
        if self._recv_reset is not None and not self._recv_buf:
            raise StreamReset(self._recv_reset)
        if not self._recv_buf:
            return b""  # clean EOF
        n = min(max_bytes, len(self._recv_buf))
        out = bytes(self._recv_buf[:n])
        del self._recv_buf[:n]
        self._mux._emit(_encode(self.id, WINDOW, n.to_bytes(4, "big")))  # replenish
        return out

    # ── mux-internal callbacks ────────────────────────────────────────
    def _on_data(self, payload: bytes) -> None:
        self._recv_buf.extend(payload)
        self._data_ready.set()

    def _on_end(self) -> None:
        self._recv_eof = True
        self._data_ready.set()

    def _on_window(self, credit: int) -> None:
        self._send_window += credit
        while self._send_waiters and self._send_window > 0:
            self._send_waiters.popleft().set_result(None)

    def _fail(self, code: int) -> None:
        self._recv_reset = code
        self._data_ready.set()
        while self._send_waiters:
            w = self._send_waiters.popleft()
            if not w.done():
                w.set_result(None)


class Mux:
    """Multiplexes :class:`Stream` s over one ordered byte channel.

    ``on_send(frame_bytes)`` ships a frame (wire it to ``transport.encrypt`` +
    the relay socket); :meth:`feed` consumes an inbound frame. ``on_stream`` (if
    given) fires when the peer opens a stream.
    """

    def __init__(self, *, is_initiator: bool, on_send, on_stream=None) -> None:
        self._on_send = on_send
        self._on_stream = on_stream
        self._streams: dict[int, Stream] = {}
        self._next_id = 1 if is_initiator else 2
        self._closed = False

    def open(self, open_info: bytes = b"") -> Stream:
        """Open a new outbound stream."""
        sid = self._next_id
        self._next_id += 2
        s = Stream(self, sid, open_info)
        self._streams[sid] = s
        self._emit(_encode(sid, OPEN, open_info))
        return s

    def feed(self, frame: bytes) -> None:
        """Dispatch one inbound frame."""
        if self._closed:
            return
        sid, ftype, payload = _decode(frame)
        if ftype == OPEN:
            s = Stream(self, sid, payload)
            self._streams[sid] = s
            if self._on_stream is not None:
                self._on_stream(s)
            return
        s = self._streams.get(sid)
        if s is None:
            return  # unknown/closed stream — ignore (peer may have raced a RESET)
        if ftype == DATA:
            s._on_data(payload)
        elif ftype == END:
            s._on_end()
        elif ftype == WINDOW:
            # WINDOW carries a u32 credit — a wrong-length payload is a malformed
            # control frame; abort the stream rather than crash or inflate credit.
            if len(payload) != 4:
                s.reset()
            else:
                s._on_window(int.from_bytes(payload, "big"))
        elif ftype == RESET:
            self._drop(sid)
            s._fail(payload[0] if payload else 0)

    def close(self) -> None:
        """Tear down every stream (e.g. the transport dropped)."""
        self._closed = True
        for s in list(self._streams.values()):
            s._fail(0)
        self._streams.clear()

    # internal
    def _emit(self, frame: bytes) -> None:
        if not self._closed:
            self._on_send(frame)

    def _drop(self, sid: int) -> None:
        self._streams.pop(sid, None)

File: web/src/homefree/mux.ts
// Browser mirror of the Home Free stream multiplexer (agent side: mux.py) — #579 P1.
//
// The E2E `Transport` (handshake.ts) is one ordered, reliable, encrypted byte
// channel. To carry the box's whole app — many concurrent `/api` requests plus
// the terminal `/ws` — over it, we multiplex independent **streams**. The mux is
// transport-agnostic: it emits opaque frame bytes via `onSend` and consumes them
// via `feed`, so the relay still only ever sees ciphertext (no relay change).
//
// Wire (must match mux.py + mux.vectors.json exactly):
//   stream_id(u32) · type(u8) · len(u24) · payload[len]
// Types: OPEN=1, DATA=2, END=3 (half-close), RESET=4 (abort), WINDOW=5 (credit).
//
// Flow control is credit-based per stream: a receiver advertises INITIAL_WINDOW
// and only replenishes (WINDOW) as its consumer reads, so an un-read stream
// drains the sender's window and applies real backpressure. Stream ids follow the
// HTTP/2 convention — the initiator opens odd ids, the responder even.

export const OPEN = 1;
export const DATA = 2;
export const END = 3;
export const RESET = 4;
export const WINDOW = 5;

export const INITIAL_WINDOW = 256 * 1024;
export const MAX_CHUNK = 16 * 1024;
const U24_MAX = (1 << 24) - 1;

export function encodeFrame(streamId: number, ftype: number, payload: Uint8Array = new Uint8Array(0)): Uint8Array {
  if (payload.length > U24_MAX) throw new Error("mux frame payload too large");
  const out = new Uint8Array(8 + payload.length);
  const dv = new DataView(out.buffer);
  dv.setUint32(0, streamId >>> 0, false);
  out[4] = ftype & 0xff;
  out[5] = (payload.length >>> 16) & 0xff;
  out[6] = (payload.length >>> 8) & 0xff;
  out[7] = payload.length & 0xff;
  out.set(payload, 8);
  return out;
}

export function decodeFrame(frame: Uint8Array): { streamId: number; ftype: number; payload: Uint8Array } {
  if (frame.length < 8) throw new Error("short mux frame");
  const dv = new DataView(frame.buffer, frame.byteOffset, frame.byteLength);
  const streamId = dv.getUint32(0, false);
  const ftype = frame[4];
  const length = (frame[5] << 16) | (frame[6] << 8) | frame[7];
  if (frame.length - 8 !== length) throw new Error("truncated mux frame");
  return { streamId, ftype, payload: frame.subarray(8, 8 + length) };
}

export class StreamReset extends Error {
  readonly code: number;
  constructor(code = 0) {
    super(`stream reset (code ${code})`);
    this.code = code;
  }
}

// A one-shot resettable notifier — the TS analogue of asyncio.Event, used to
// wake a pending read()/write() when data/window/eof/reset arrives.
class Gate {
  private _p!: Promise<void>;
  private _res!: () => void;
  constructor() {
    this.reset();
  }
  reset(): void {
    this._p = new Promise((r) => (this._res = r));
  }
  wait(): Promise<void> {
    return this._p;
  }
  set(): void {
    this._res();
  }
}

export class Stream {
  readonly id: number;
  readonly openInfo: Uint8Array;
  private _mux: Mux;
  private _sendWindow: number;
  private _sendWaiters: Array<() => void> = [];
  private _sendEnded = false;
  private _recvBuf: number[] = [];
  private _recvEof = false;
  private _recvReset: number | null = null;
  private _dataReady = new Gate();

  constructor(mux: Mux, id: number, openInfo: Uint8Array) {
    this._mux = mux;
    this.id = id;
    this.openInfo = openInfo;
    this._sendWindow = mux.initialWindow;
  }

  async write(data: Uint8Array): Promise<void> {
    if (this._sendEnded) throw new StreamReset(0);
    let off = 0;
    while (off < data.length) {
      while (this._sendWindow <= 0) {
        if (this._recvReset !== null) throw new StreamReset(this._recvReset);
        await new Promise<void>((r) => this._sendWaiters.push(r));
      }
      const n = Math.min(data.length - off, this._sendWindow, this._mux.maxChunk);
      this._mux._emit(encodeFrame(this.id, DATA, data.subarray(off, off + n)));
      this._sendWindow -= n;
      off += n;
    }
  }

  async end(): Promise<void> {
    if (!this._sendEnded) {
      this._sendEnded = true;
      this._mux._emit(encodeFrame(this.id, END));
    }
  }

  reset(code = 0): void {
    this._mux._emit(encodeFrame(this.id, RESET, new Uint8Array([code & 0xff])));
    this._mux._drop(this.id);
    this._fail(code);
  }

  /** Up to `maxBytes` of received data; empty array at clean EOF. Reading replenishes the peer window. */
  async read(maxBytes = MAX_CHUNK): Promise<Uint8Array> {
    while (this._recvBuf.length === 0 && !this._recvEof && this._recvReset === null) {
      this._dataReady.reset();
      await this._dataReady.wait();
    }
    if (this._recvReset !== null && this._recvBuf.length === 0) throw new StreamReset(this._recvReset);
    if (this._recvBuf.length === 0) return new Uint8Array(0); // clean EOF
    const n = Math.min(maxBytes, this._recvBuf.length);
    const out = Uint8Array.from(this._recvBuf.splice(0, n));
    const credit = new Uint8Array(4);
    new DataView(credit.buffer).setUint32(0, n, false);
    this._mux._emit(encodeFrame(this.id, WINDOW, credit)); // replenish
    return out;
  }

  _onData(payload: Uint8Array): void {
    for (const b of payload) this._recvBuf.push(b);
    this._dataReady.set();
  }
  _onEnd(): void {
    this._recvEof = true;
    this._dataReady.set();
  }
  _onWindow(credit: number): void {
    this._sendWindow += credit;
    while (this._sendWaiters.length && this._sendWindow > 0) this._sendWaiters.shift()!();
  }
  _fail(code: number): void {
    this._recvReset = code;
    this._dataReady.set();
    while (this._sendWaiters.length) this._sendWaiters.shift()!();
  }
}

export interface MuxOptions {
  isInitiator: boolean;
  onSend: (frame: Uint8Array) => void;
  onStream?: (s: Stream) => void;
  initialWindow?: number;
  maxChunk?: number;
}

export class Mux {
  readonly initialWindow: number;
  readonly maxChunk: number;
  private _onSend: (f: Uint8Array) => void;
  private _onStream?: (s: Stream) => void;
  private _streams = new Map<number, Stream>();
  private _nextId: number;
  private _closed = false;

  constructor(opts: MuxOptions) {
    this._onSend = opts.onSend;
    this._onStream = opts.onStream;
    this._nextId = opts.isInitiator ? 1 : 2;
    this.initialWindow = opts.initialWindow ?? INITIAL_WINDOW;
    this.maxChunk = opts.maxChunk ?? MAX_CHUNK;
  }

  open(openInfo: Uint8Array = new Uint8Array(0)): Stream {
    const sid = this._nextId;
    this._nextId += 2;
    const s = new Stream(this, sid, openInfo);
    this._streams.set(sid, s);
    this._emit(encodeFrame(sid, OPEN, openInfo));
    return s;
  }

  feed(frame: Uint8Array): void {
    if (this._closed) return;
    const { streamId, ftype, payload } = decodeFrame(frame);
    if (ftype === OPEN) {
      const s = new Stream(this, streamId, payload.slice());
      this._streams.set(streamId, s);
      this._onStream?.(s);
      return;
    }
    const s = this._streams.get(streamId);
    if (!s) return; // unknown/closed stream — ignore
    if (ftype === DATA) s._onData(payload);
    else if (ftype === END) s._onEnd();
    else if (ftype === WINDOW) {
      // WINDOW carries a u32 credit — a wrong-length payload is a malformed
      // control frame; abort the stream rather than throw or inflate credit.
      if (payload.length !== 4) s.reset();
      else s._onWindow(new DataView(payload.buffer, payload.byteOffset, payload.byteLength).getUint32(0, false));
    } else if (ftype === RESET) {
      this._drop(streamId);
      s._fail(payload.length ? payload[0] : 0);
    }
  }

  close(): void {
    this._closed = true;
    for (const s of this._streams.values()) s._fail(0);
    this._streams.clear();
  }

  _emit(frame: Uint8Array): void {
    if (!this._closed) this._onSend(frame);
  }
  _drop(sid: number): void {
    this._streams.delete(sid);
  }
}
